aard.1
.\"MIT/X Consortium License
.\"
.\"© 2021-2022 Carl H. Henriksson <aard at martial dot cc>
.\"
.\"Permission is hereby granted, free of charge, to any person obtaining a
.\"copy of this software and associated documentation files (the "Software"),
.\"to deal in the Software without restriction, including without limitation
.\"the rights to use, copy, modify, merge, publish, distribute, sublicense,
.\"and/or sell copies of the Software, and to permit persons to whom the
.\"Software is furnished to do so, subject to the following conditions:
.\"
.\"The above copyright notice and this permission notice shall be included in
.\"all copies or substantial portions of the Software.
.\"
.\"THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\"IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\"FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
.\"THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\"LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
.\"FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
.\"DEALINGS IN THE SOFTWARE.
.Dd $Mdocdate$
.Dt AARD 1
.Os
.Sh NAME
.Nm aard
.Nd modular action interface
.Sh SYNOPSIS
.Nm
.Op Fl v
.Op Ar option ...
.Sh DESCRIPTION
The
.Nm
utility facilitates the execution of arbitrary actions in X.
.Pp
Enabling the user to execute multiple functions (options) in series,
.Nm
provides a modular interface to action composition.
.Pp
The program uses an internal <buffer> to hold data. That <buffer> can read from,
or write to the X clipboard.
By default, the <buffer> is written to the X clipboard at succesful program exit.
.Pp
The options are provided as arguments.
Notably, every option can be destructive to the data in the <buffer>,
overwriting or modifying it. The options are as follows:
.Bl -tag -width Ds
.It Fl v
Print version information and exit program.
.It Ar -
Populate the <buffer> with data read from standard input.
.It Ar base64
Encode the data in the <buffer> to base64 and load that base64 data into the <buffer>.
.It Ar clean
Remove leading and trailing whitespace, and newline characters from the <buffer>.
.It Ar clear
Remove all data in the <buffer>.
.It Ar define
Fetch definition of the grammatical unit currently in the <buffer> from
an online source. By default, a pdf with the definition is created and opened,
and the filesystem path to that pdf is loaded into the <buffer>.
.It Ar edit
Edit the data in the <buffer> with a text editor. This requires
.Nm
to be called from a terminal.
.It Ar emoji
Select an emoji from a prompt, and load that emoji into the <buffer>.
.It Ar fetch
Select an entity from the filesystem, the
.Nm
log, or the users shell history. Load the selected entity into the <buffer>.
.It Ar file
Interpret the <buffer> data as a filesystem path, and load the content of that
file into the <buffer>.
.It Ar help
List all
.Nm
options and their explanations in a prompt, enabling the user to select one option.
That option is then executed.
.It Ar host
List contents of the system hosts file in a prompt, enabling the user to select one entry.
The ip of the selected entry is then loaded into the <buffer>.
.It Ar html
Interpret the data in the <buffer> as an URL, and request the html source from that URL.
That target source is then loaded into the <buffer>.
.It Ar htmlx
Interpret the data in the <buffer> as an URL, and request the html render from that URL.
That text render is then loaded into the <buffer>.
.It Ar image
Interpret the data in the <buffer> as a filesystem path, and open that path with an image viewer.
.It Ar light
Interpret the data in the <buffer> as an integer representing percent.
Set the X display backlight to that percent.
.It Ar log
Timestamp and append the data in the <buffer> to the
.Nm
log.
.It Ar media
Interpret the data in the <buffer> as a filesystem path, and open that path with a media player.
.It Ar option
Open a prompt, enabling the user to input
.Nm
options. Then execute
.Nm
with those options.
.It Ar paste
Write the data in the <buffer> to standard output.
.It Ar pdf
Create a pdf with the data in the <buffer>, then load the filesystem path to that
created pdf into the <buffer>.
Default behaviour is to open that pdf with a pdf viewer after creation.
.It Ar post
Open a prompt, enabling the user to input text. That text is then postfixed to the
data in the <buffer>.
.It Ar pre
Open a prompt, enabling the user to input text. That text is then prefixed to the
data in the <buffer>.
.It Ar qr
Create a QR code with the data in the <buffer> and save it as png,
then load the filesystem path to that created png file into the <buffer>.
Default behaviour is to open that png with an image viewer after creation.
.It Ar random
Fetch true random data from an online source. Load that data as an ASCII
encoded hexadecimal string into the <buffer>.
.It Ar report
Pass the data in the <buffer> to the reporting system.
.It Ar run
Open a prompt, enabling the user to input text. That text is then executed in a shell
as a command. The standard output, and standard error of that command, is then loaded into the <buffer>.
.It Ar select
Display the data in the <buffer> in a prompt, enabling the user to select one entry.
That entry is loaded into the <buffer>.
.It Ar set
Open a prompt, enabling the user to input text. That text is loaded into the <buffer>.
.It Ar tty
Open a terminal, as defined in the
.Nm
configuration file.
.It Ar url
Run an URL isolation program with the data in the <buffer>, generating a list of the URL addresses found
in the <buffer>. Then list those URL's in a prompt, enabling the user to select one URL.
That URL is loaded into the <buffer>.
.It Ar word
Select a dictionary word from a prompt, and load it into the <buffer>.
.It Ar x
Interpret the data in the <buffer> as a command. That command is executed in a shell.
The operations standard output, and standard error, is be loaded into the <buffer>.
.It Ar yt
Interpret the data in the <buffer> as YouTube search terms, and request search results for those terms online.
Then list the returned results in a prompt, enabling the user to select one entity.
The URL to that entity is loaded into the <buffer>.
.El
.Sh CONFIGURATION
.Bl -tag -width Ds
.It Ev AARD_CLIP_GET
Command for reading the X clipboard
.It Ev AARD_CLIP_SET
Command for setting the X clipboard
.It Ev AARD_DATE_FORMAT
Format definition for time and date printing.
.It Ev AARD_DICTIONARY_PATH
Path to word list file used in: word
.It Ev AARD_EDITOR
Command for editing text files used in: editor
.It Ev AARD_EMOJI_PATH
Path to emoji list file used in: emoji
.It Ev AARD_HISTORY_PATH
Path to shell history file used in: fetch
.It Ev AARD_IMAGE
Command for viewing image files used in: image qr
.It Ev AARD_LOG_PATH
Path to log file used in: fetch log media yt
.It Ev AARD_MEDIA
Command for playing media files used in: media
.It Ev AARD_PDF
Command for viewing pdf files used in: define pdf
.It Ev AARD_PDF_PATH
Path to pdf file used in: pdf
.It Ev AARD_PROMPT
Command for selecting single entries from lists
.It Ev AARD_QR_PATH
Path to QR image file used in: qr
.It Ev AARD_ROOT
aard configuration directory. Default path is:
.Pa $HOME/.config/aard
.It Ev AARD_TTY
Command for opening terminal, used in: tty
.It Ev AARD_X_CLIPBOARD
Configuration option defining whether to write the <buffer> to the
X clipboard on succesful program exit. Leave empty for "no".
.It Ev AARD_X_DEFINEPDF
Configuration option defining whether to create a pdf with the
definition text or not. Leave empty for "no".
.It Ev AARD_X_PDFSHOW
Configuration option defining whether to open the created pdf in the pdf viewer
defined in AARD_PDF or not. Leave empty for "no".
.It Ev AARD_X_REPORT
Configuration option defining whether to use the reporting
system or not. Leave empty for "no"
.It Ev AARD_X_STATUS
Configuration option defining whether to use the status
system or not. Leave empty for "no"
.El
.Sh FILES
.Bl -tag -width Ds
.It Pa $AARD_ROOT/aard.conf
Necessary file containing aliases, paths and configuration.
.El
.Sh EXIT STATUS
.Ex -std
.Sh EXAMPLES
List all options, allowing the user to select one and run
.Nm
with that option
.Pp
.Dl aard help
.Pp
Find spelling of a word, and request the definition of that word online
.Pp
.Dl aard word define
.Pp
Find and select emoji based on keywords, and copy it to clipboard
.Pp
.Dl aard emoji
.Pp
Find file, select line from it, and copy that line into clipboard.
.Pp
.Dl aard fetch file select
.Pp
Fetch command from shell history, edit it, then execute it (the edit
option must be executed from a terminal)
.Pp
.Dl aard fetch edit x
.Sh DIAGNOSTICS
.Nm
includes a reporting system for descriptive error messages,
and a simpler status reporting system for program state.
Example scripts are installed as aard_report and aard_status.
.Sh AUTHORS
Carl H. Henriksson <aard at martial dot cc>
.Sh CAVEATS
The aard_report example script updates the name of the root window.
Disabled the system in the configuration file if that behaviour is unwanted.
.Sh SECURITY CONSIDERATIONS
.Nm
uses temporary files. Those files containing potentially sensitive
data may be left in the filesystem on program fail.
.El


aard.sh
#!/bin/sh
#
#	aard-VERSION
#
# MIT/X Consortium License
#
# © 2021-2022 Carl H. Henriksson <aard at martial dot cc>
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

AARD_OPTION="\
base64	b	encode the data in <buffer> to base64
clean	cn	clean up the data in <buffer>
clear	cr	remove all data in <buffer>
define	d	fetch definition of grammatical unit in <buffer>
edit	e	edit the data in <buffer> with a text editor
emoji	em	select and emoji and load it into <buffer>
fetch	f	load filesystem, log, or shell history entity into <buffer>
file	fi	read the file data from path in <buffer>
help	h	list all options and select one for execution
host	ho	select and load ip from the system hosts into <buffer>
html	ht	load source from url to <buffer>
htmlx	hx	dump render from url to <buffer>
image	i	open the <buffer> as an image file
light	li	set X backlight to percent integer value in <buffer>
log	l	timestamp and append the <buffer> to the log file
media 	m	open the <buffer> as a media file
option	o	input options and execute them
paste	p	write the <buffer> to standard output
pdf	pd	create a pdf with the data in <buffer>
post	po	input and add postfix to <buffer>
pre	pr	input and add prefix to <buffer>
qr	q	create a QR code png with the data in <buffer>
random	ra	load random ascii encoded hex into <buffer>
report	re	pass the <buffer> to the report system
run	r	input and execute commands in <buffer> loading output to <buffer>
select	se	isolate one line in the <buffer>
set	s	set <buffer>
stdin	-	populate <buffer> with data read from standard input
tty	t	open terminal
url	u	isolate and select an url in <buffer>
word	w	select dictionary word and load it into <buffer>
x	x	execute <buffer> as command loading output to <buffer>
yt	y	search and select youtube video url"

# Error and system interface

aard_log() {
	printf '%s %s\n' "$(aard_date)" "$(cat "$AARD_BUFFER")" >> "$AARD_LOG_PATH" \
		|| aard_quit 1 'aard_log: Failed to write log'
}

aard_msg() {
	aard_report_set "$(aard_date) aard: $*" || exit 1
}

aard_quit() {
	if [ $# -eq 2 ]; then
		aard_msg "$2"
	fi

	aard_status_set 'FAIL'
	AARD_QUIT_RETURN='1'

	if [ $# -eq 1 ]; then
		AARD_QUIT_RETURN="$1"
	fi

	aard_post

	exit "$AARD_QUIT_RETURN"
}

aard_report_set() {
	if [ -z "$AARD_X_REPORT" ]; then
		return 0
	fi

	command -v aard_report > /dev/null 2>&1 || return 1
	aard_report "$*"
}

aard_status_set() {
	if [ -z "$AARD_X_STATUS" ]; then
		return 0
	fi

	command -v aard_status > /dev/null 2>&1 || return 1
	aard_status "$*"
}

# Utilities

aard_date() {
	date +"$AARD_DATE_FORMAT" 2> /dev/null || aard_quit 1 'aard_date: Failed to get date'
}

# Clipboard

aard_clip_get() {
	eval " $AARD_CLIP_GET" 2> /dev/null || aard_quit 1 'aard_clip_get: Failed to read clipboard'
}

aard_clip_set() {
	eval " $AARD_CLIP_SET" 2> /dev/null || aard_quit 1 'aard_clip_set: Failed to write clipboard'
}

# Etc

aard_prompt() {
	if [ -z "$AARD_PROMPT" ]; then
		aard_quit 1 'aard_prompt: Failed to read variable: AARD_PROMPT'
	fi

	if [ $# -ne 1 ]; then
		AARD_PROMPT_MSG='aard:'
	else
		AARD_PROMPT_MSG="$1"
	fi

	AARD_PROMPT_CMD="$AARD_PROMPT \"$AARD_PROMPT_MSG\""

	eval " $AARD_PROMPT_CMD" 2> /dev/null || aard_quit 2 'aard_prompt: Failed to open prompt'
}

aard_select() {
	aard_prompt 'aard select:' || aard_quit 1 'aard_select: Failed to select line'
}

# Main

aard_arg() {
	if [ $# -eq 0 ]; then
		printf 'aard: No arguments provided\n'
		exit 1
	fi

	case "$1" in
	--help)
		printf '%s\n' 'usage: aard [-v] [option ...]'
		;;
	-v)
		printf '%s\n' 'aard-VERSION'
		;;
	*)
		return
		;;
	esac

	exit 0
}

aard_iterate() {
	for option in "$@"; do
		aard_process "$option"
	done
}

aard_process() {
	if [ $# -ne 1 ]; then
		aard_quit 1 'aard_process: Failed to provide valid argument amount'
	fi

	case "$1" in
	base64 | b)
		command -v b64encode > /dev/null 2>&1 \
			|| aard_quit 1 'aard_process base64: Failed to find program: b64encode'

		b64encode aard < "$AARD_BUFFER" > "$AARD_FILE" 2> /dev/null \
			|| aard_quit 2 'aard_process base64: Failed to encode base64'

		cat "$AARD_FILE" > "$AARD_BUFFER"
		;;
	clean | cn)
		AARD_CLEAN_WHITESPACE="$(awk '$1=$1' "$AARD_BUFFER" 2> /dev/null \
			|| aard_quit 1 'aard_process clean: Failed to clean whitespace')"

		printf '%s' "$AARD_CLEAN_WHITESPACE" | tr -d '\n' > "$AARD_BUFFER" 2> /dev/null \
			|| aard_quit 2 'aard_process clean: Failed to clean newline'
		;;
	clear | cr)
		printf '' > "$AARD_BUFFER"
		;;
	define | d)
		command -v w3m > /dev/null 2>&1 || aard_quit 1 'aard_process define: Failed to find program: w3m'

		AARD_DEFINE_WORD="$(cat "$AARD_BUFFER")"
		AARD_DEFINE_URL="http://merriam-webster.com/dictionary/$AARD_DEFINE_WORD"

		w3m -config /dev/null -T html/text "$AARD_DEFINE_URL" > "$AARD_FILE" 2> /dev/null \
			|| aard_quit 2 'aard_process define: Failed to fetch html'

		AARD_FORMAT="$(awk '/Definition of/,/Synonyms for*/;/Synonyms for/{exit}' "$AARD_FILE" 2> /dev/null \
			|| aard_quit 3 'aard_process define: Failed to format html')"

		AARD_CLEAN="$(printf '%s' "$AARD_FORMAT" | sed '$d' 2> /dev/null \
			|| aard_quit 4 'aard_process define: Failed to clean data')"

		printf '%s %s\n' "$AARD_DEFINE_WORD" "$AARD_CLEAN" > "$AARD_BUFFER"

		if [ -n "$AARD_X_DEFINEPDF" ]; then
			aard_process 'pdf'
		fi
		;;
	edit | e)
		cat "$AARD_BUFFER" > "$AARD_FILE"

		eval " $AARD_EDITOR $AARD_FILE" \
			|| aard_quit 1 'aard_process edit: Failed to edit temporary file'

		cat "$AARD_FILE" > "$AARD_BUFFER"
		;;
	emoji | em)
		if [ -z "$AARD_EMOJI_PATH" ]; then
			aard_quit 1 'aard_process emoji: Failed to read variable: AARD_EMOJI_PATH'
		fi

		awk -F'<[^>]*>' '
			NR>25 && NR<31506 && /chars/ && length($2) > 1 {
				printf $2 "\t\t"
			} /class=\047name/ {
				gsub(/\&amp;/, "and");
				if (length($5) < 7)
					print $2;
					else print $5
			}' OFS='' "$AARD_EMOJI_PATH" > "$AARD_FILE" 2> /dev/null \
			|| aard_quit 2 'aard_process emoji: Failed to format emoji list'

		AARD_EMOJI_CHOSEN="$(expand "$AARD_FILE" | aard_select 'aard emoji')"

		AARD_EMOJI_ISOLATED="$(printf '%s' "$AARD_EMOJI_CHOSEN" | awk -F' ' '{print $1}' 2> /dev/null \
			|| aard_quit 3 'aard_process emoji: Failed to isolate emoji')"

		printf '%s' "$AARD_EMOJI_ISOLATED" > "$AARD_BUFFER"
		;;
	fetch | f)
		find "$HOME" > "$AARD_FILE" 2> /dev/null

		if [ -n "$AARD_HISTORY_PATH" ]; then
			tail -r "$AARD_HISTORY_PATH" >> "$AARD_FILE" 2> /dev/null
		fi

		if [ -n "$AARD_LOG_PATH" ]; then
			tail -r "$AARD_LOG_PATH" >> "$AARD_FILE" 2> /dev/null
		fi

		aard_select 'aard fetch' > "$AARD_BUFFER" < "$AARD_FILE"
		;;
	file | fi)
		AARD_FILENAME="$(cat "$AARD_BUFFER")"
		if [ ! -r "$AARD_FILENAME" ]; then
			aard_quit 1 'aard_process file: Failed to read file'
		fi
		cat "$AARD_FILENAME" > "$AARD_BUFFER"
		;;
	help | h)
		printf '%s' "$AARD_OPTION" | expand > "$AARD_FILE" 2> /dev/null \
			|| aard_quit 1 'aard_process help: Failed to expand help'

		AARD_HELP_SELECTION="$(aard_select 'aard help' < "$AARD_FILE" | awk '{print $1}' 2> /dev/null \
			|| aard_quit 2 'aard_process help: Failed to read selection')"

		aard_process "$AARD_HELP_SELECTION"
		;;
	host | ho)
		AARD_HOST_EXPANDED="$(expand /etc/hosts 2> /dev/null | aard_select 'aard host' \
			|| aard_quit 1 'aard_process host: Failed to expand selection')"

		AARD_HOST_SELECTION="$(printf '%s' "$AARD_HOST_EXPANDED" | awk '{print $1}' 2> /dev/null \
			|| aard_quit 2 'aard_process host: Failed to isolate selection')"

		printf '%s' "$AARD_HOST_SELECTION" > "$AARD_BUFFER"
		;;
	html | ht)
		command -v w3m > /dev/null 2>&1 || aard_quit 1 'aard_process html: Failed to find program: w3m'

		w3m -config /dev/null -dump_source "$(cat "$AARD_BUFFER")" > "$AARD_FILE" 2> /dev/null \
			|| aard_quit 2 'aard_process html: Failed to fetch html'

		cat "$AARD_FILE" > "$AARD_BUFFER"
		;;
	htmlx | hx)
		command -v w3m > /dev/null 2>&1 || aard_quit 1 'aard_process htmlx: Failed to find program: w3m'

		w3m -config /dev/null -T html/text "$(cat "$AARD_BUFFER")" > "$AARD_FILE" 2> /dev/null \
			|| aard_quit 2 'aard_process htmlx: Failed to fetch html'

		cat "$AARD_FILE" > "$AARD_BUFFER"
		;;
	image | i)
		if [ -z "$AARD_IMAGE" ]; then
			aard_quit 1 'aard_process image: Failed to read variable: AARD_IMAGE'
		fi

		eval " $AARD_IMAGE $(cat "$AARD_BUFFER")" 2> /dev/null \
			|| aard_quit 2 'aard_process image: Failed to show image'
		;;
	light | li)
		command -v xbacklight > /dev/null 2>&1 \
			|| aard_quit 1 'aard_process light: Failed to find program: xbacklight'

		xbacklight -set "$(cat "$AARD_BUFFER")" 2> /dev/null \
			|| aard_quit 2 'aard_process light: Failed to set light'
		;;
	log | l)
		aard_log
		;;
	media | m)
		if [ -z "$AARD_MEDIA" ]; then
			aard_quit 1 'aard_process media: Failed to read variable: AARD_MEDIA'
		fi

		AARD_TARGET="$(cat "$AARD_BUFFER")"

		eval " $AARD_MEDIA $AARD_TARGET" 2> /dev/null \
			|| aard_quit 2 'aard_process media: Failed to play media'
		;;
	option | o)
		aard_iterate $(printf '' | aard_prompt 'aard option:')
		;;
	paste | p)
		cat "$AARD_BUFFER"
		;;
	pdf | pd)
		command -v enscript > /dev/null 2>&1 \
			|| aard_quit 1 'aard_process pdf: Failed to find program: enscript'

		command -v ps2pdf > /dev/null 2>&1 \
			|| aard_quit 2 'aard_process pdf: Failed to find program: ps2pdf'

		if [ -z "$AARD_PDF" ]; then
			aard_quit 3 'aard_process pdf: Failed to read variable: AARD_PDF'
		fi

		if [ -z "$AARD_PDF_PATH" ]; then
			aard_quit 4 'aard_process pdf: Failed to read variable: AARD_PDF_PATH'
		fi

		enscript -p - > "$AARD_FILE" 2> /dev/null < "$AARD_BUFFER" \
			|| aard_quit 5 'aard_process pdf: Failed to generate postscript'

		ps2pdf "$AARD_FILE" "$AARD_PDF_PATH" 2> /dev/null \
			|| aard_quit 6 'aard_process pdf: Failed to generate pdf'

		printf '%s' "$AARD_PDF_PATH" > "$AARD_BUFFER"

		if [ -n "$AARD_X_PDFSHOW" ]; then
			AARD_TARGET="$(cat "$AARD_BUFFER")"

			eval " $AARD_PDF $AARD_TARGET" 2> /dev/null \
				|| aard_quit 7 'aard_process pdf: Failed to show pdf'
		fi
		;;
	post | po)
		AARD_POSTFIX="$(printf '' | aard_prompt 'post:' \
			|| aard_quit 1 'aard_process post: Failed to read user input')"

		printf '%s%s' "$(cat "$AARD_BUFFER")" "$AARD_POSTFIX" > "$AARD_BUFFER"
		;;
	pre | pr)
		AARD_PREFIX="$(printf '' | aard_prompt 'pre:' \
			|| aard_quit 1 'aard_process pre: Failed to read user input')"

		printf '%s%s' "$AARD_PREFIX" "$(cat "$AARD_BUFFER")" > "$AARD_BUFFER"
		;;
	qr | q)
		command -v qrencode > /dev/null 2>&1 \
			|| aard_quit 1 'aard_process qr: Failed to find program: qrencode'

		if [ -z "$AARD_QR_PATH" ]; then
			aard_quit 2 'aard_process qr: Failed to read variable: AARD_QR_PATH'
		fi

		qrencode -s 24 -o "$AARD_QR_PATH" 2> /dev/null < "$AARD_BUFFER" \
			|| aard_quit 3 'aard_process qr: Failed to generate QR'

		printf '%s' "$AARD_QR_PATH" > "$AARD_BUFFER"

		aard_process image
		;;
	random | ra)
		command -v w3m > /dev/null 2>&1 \
			|| aard_quit 1 'aard_process random: Failed to find program: w3m'

		w3m -dump_source 'http://random.org/cgi-bin/randbyte?nbytes=4' > "$AARD_FILE" 2> /dev/null \
			|| aard_quit 2 'aard_process random: Failed to fetch html'

		AARD_RANDOM_ISOLATED="$(hexdump -n 8 -e '2/4 "%08X"' "$AARD_FILE" 2> /dev/null \
			|| aard_quit 3 'aard_process random: Failed to format hex')"

		printf '%s' "$AARD_RANDOM_ISOLATED" > "$AARD_BUFFER"

		aard_process clean
		;;
	report | re)
		aard_report_set "$(cat "$AARD_BUFFER")" \
			|| aard_quit 1 'aard_process report: Failed to report'
		;;
	run | r)
		aard_iterate 'set' 'x'
		;;
	select | se)
		aard_select 'aard select' > "$AARD_FILE" < "$AARD_BUFFER"
		cat "$AARD_FILE" > "$AARD_BUFFER"
		;;
	set | s)
		AARD_SET_INPUT="$(printf '' | aard_prompt 'aard set:' \
			|| aard_quit 1 'aard_process set: Failed to read user input')"

		printf '%s' "$AARD_SET_INPUT" > "$AARD_BUFFER"
		;;
	stdin | -)
		cat - > "$AARD_BUFFER"
		;;
	tty | t)
		if [ -z "$AARD_TTY" ]; then
			aard_quit 1 'aard_process tty: Failed to read variable: AARD_TTY'
		fi

		eval " $AARD_TTY" 2> /dev/null \
			|| aard_quit 2 'aard_process tty: Failed to open terminal'
		;;
	url | u)
		command -v urlisolator > /dev/null 2>&1 \
			|| aard_quit 1 'aard_process url: Failed to find program: urlisolator'

		urlisolator > "$AARD_FILE" 2> /dev/null < "$AARD_BUFFER" \
			|| aard_quit 2 'aard_process url: Failed to isolate url'

		aard_select 'aard url' > "$AARD_BUFFER" < "$AARD_FILE"
		;;
	word | w)
		if [ -z "$AARD_DICTIONARY_PATH" ]; then
			aard_quit 1 'aard_process word: Failed to read variable: AARD_DICTIONARY_PATH'
		fi

		tr -d '\r' < "$AARD_DICTIONARY_PATH" > "$AARD_FILE" 2> /dev/null \
			|| aard_quit 2 'aard_process word: Failed to fix newline'

		aard_select 'aard word' > "$AARD_BUFFER" < "$AARD_FILE"
		;;
	x)
		cat "$AARD_BUFFER" > "$AARD_FILE"

		chmod +x "$AARD_FILE" 2> /dev/null \
			|| aard_quit 1 'aard_process x: Failed to change file permissions'

		AARD_X_OUTPUT="$(sh "$AARD_FILE" 2>&1)"

		printf '%s' "$AARD_X_OUTPUT" > "$AARD_BUFFER"
		;;
	yt | y)
		command -v yt-dlp > /dev/null 2>&1 \
			|| aard_quit 1 'aard_process yt: Failed to find program: yt-dlp'

		yt-dlp -j "ytsearch16:$(cat "$AARD_BUFFER")" > "$AARD_FILE" 2> /dev/null \
			|| aard_quit 2 'aard_process yt: Failed to find URLs'

		AARD_YT_SEARCH="$(awk -F'"' '{print $4, $8}' "$AARD_FILE" 2> /dev/null)" \
			|| aard_quit 3 'aard_process yt: Failed to format URLs'

		AARD_YT_URL="$(printf '%s' "$AARD_YT_SEARCH" | aard_select 'aard yt' \
		| awk '{print "http://youtube.com/watch?v=" $1}' 2> /dev/null \
			|| aard_quit 4 'aard_process yt: Failed to format URL')"

		printf '%s' "$AARD_YT_URL" > "$AARD_BUFFER"

		aard_log
		;;
	*)
		aard_quit 1 "aard_process *: Failed to identify argument; $*"
		;;
	esac
}

aard_post() {
	rm "$AARD_BUFFER" 2> /dev/null
	rm "$AARD_FILE" 2> /dev/null
}

aard_pre() {
	# Configuration
	if [ -z "$AARD_CONF" ]; then
		# This variable is defined in the Makefile
		AARD_CONF="MAKECONFROOT/aard.conf"
		if [ -z "$AARD_CONF" ]; then
			aard_quit 1 'aard_run: Failed to find configuration file'
		fi
	fi

	if [ ! -r "$AARD_CONF" ]; then
		aard_quit 2 'aard_run: Failed to read file'
	fi

	. "$AARD_CONF"

	# Files
	AARD_BUFFER="$(mktemp -t aard."$$".XXXXXXXX 2> /dev/null \
		|| aard_quit 3 'aard_run: Failed to create buffer file')"
	AARD_FILE="$(mktemp -t aard."$$".XXXXXXXX 2> /dev/null \
		|| aard_quit 4 'aard_run: Failed to create temporary file')"
}

aard_run() {
	aard_arg "$@"
	aard_pre

	aard_status_set 'PROCESSING'

	aard_clip_get > "$AARD_BUFFER"
	aard_iterate "$@"
	if [ -n "$AARD_X_CLIPBOARD" ]; then
		aard_clip_set < "$AARD_BUFFER"
	fi

	aard_status_set 'SUCCESS'

	aard_post
}

aard_run "$@"
